contracts/src/PBHVerifierImplV1.sol

```diff
@@ -0,0 +1,205 @@
+    /// @custom:reverts string If called more than once at the same initialisation number.
+    /// @custom:reverts InvalidWorldId if `_worldId` is set to the zero address
+    function initialize(IWorldIDGroups _worldId, uint8 _numPbhPerMonth) public reinitializer(1) {
+        // First, ensure that all of the parent contracts are initialised.
+        __delegateInit();
+
+        worldId = _worldId;
+        numPbhPerMonth = _numPbhPerMonth;
+
+        // Say that the contract is initialized.
+        __setInitialized();
+
+        emit PBHVerifierImplInitialized(_worldId, _numPbhPerMonth);
+    }
+
+    /// @notice Responsible for initialising all of the supertypes of this contract.
+    /// @dev Must be called exactly once.
+    /// @dev When adding new superclasses, ensure that any initialization that they need to perform
+    ///      is accounted for here.
+    ///
+    /// @custom:reverts string If called more than once.
+    function __delegateInit() internal virtual onlyInitializing {
+        __WorldIDImpl_init();
+    }
+
+    ///////////////////////////////////////////////////////////////////////////////
+    ///                                  Functions                             ///
+    //////////////////////////////////////////////////////////////////////////////
+
+    /// @param root The root of the Merkle tree that this proof is valid for.
+    /// @param sender The sender of this particular transaction or UserOp.
+    /// @param nonce Transaction/UserOp nonce.
+    /// @param callData Transaction/UserOp call data.
+    /// @param pbhExternalNullifier External nullifier encodeing month, year, and a pbhNonce.
+    /// @param nullifierHash Nullifier hash for this semaphore proof.
+    /// @param proof The zero-knowledge proof that demonstrates the claimer is registered with World ID.
+    function verifyPbhProof(
+        uint256 root,
+        address sender,
+        uint256 nonce,
+        bytes memory callData,
+        uint256 pbhExternalNullifier,
+        uint256 nullifierHash,
+        uint256[8] memory proof
+    ) external virtual onlyProxy onlyInitialized {
+        // First, we make sure this nullifier has not been used before.
+        if (nullifierHashes[nullifierHash]) revert InvalidNullifier();
+
+        // Verify the external nullifier
+        PBHExternalNullifier.verify(pbhExternalNullifier, numPbhPerMonth);
+        
+        // If worldId address is set, proceed with on chain verification,
+        // otherwise assume verification has been done off chain by the builder.
+        if (address(worldId) != address(0)) {
+            // We now generate the signal hash from the sender, nonce, and calldata
+            uint256 signalHash = abi.encodePacked(sender, nonce, callData).hashToField();
+
+            // We now verify the provided proof is valid and the user is verified by World ID
+            worldId.verifyProof(root, GROUP_ID, signalHash, nullifierHash, pbhExternalNullifier, proof);
+        }
```

Comment:
Do we anticipate verifying onchain? If not we should remove this logic.

Reply:
I'm not sure to be honest. Do you think that's ever something we'll need?